src/nccl_ofi_rdma.c

```diff
@@ -1450,70 +1515,73 @@ static inline int process_completions(struct fi_cq_tagged_entry *cq_entry,
-	case NCCL_OFI_RDMA_SEND_CTRL:
-		return "SEND_CTRL";
-	case NCCL_OFI_RDMA_RECV_SEGMS:
-		return "RECV_SEGMS";
-	case NCCL_OFI_RDMA_BOUNCE:
-		return "BOUNCE";
-	case NCCL_OFI_RDMA_FLUSH:
-		return "FLUSH";
-	case NCCL_OFI_RDMA_EAGER_COPY:
-		return "EAGER_COPY";
-	}
-	return "unknown";
-}
-
 /*
- * @brief	Print NCCL OFI request information
+ * @brief	Process error completion entries from the CQ error queue
+ *
+ * @return	0, on success
+ *		error, on others
  */
-static const char *nccl_net_ofi_req_str(nccl_net_ofi_rdma_req_t *req)
+static inline int process_err_completion(nccl_net_ofi_rdma_ep_t *ep,
+					 int rail_id)
 {
-	static char buf[256];
-	snprintf(buf, sizeof(buf), "{ dev: %d, size: %zu, state: %s, type: %s }",
-		 req->dev_id,
-		 req->size,
-		 req_state_str(req->state),
-		 req_type_str(req->type)
-		);
-	return buf;
-}
+	nccl_net_ofi_ep_rail_t *rail = get_rail(ep, rail_id);
+	struct fi_cq_err_entry err_entry = { 0 };
+	nccl_net_ofi_rdma_req_t *req = NULL;
+	int ret = 0;
 
-static int post_rdma_ctrl(nccl_net_ofi_rdma_req_t *req);
+	ret = fi_cq_readerr(rail->cq, &err_entry, 0);
+	if (OFI_UNLIKELY(ret == -FI_EAGAIN)) {
+		/*
+		 * Error not available yet.
+		 * fi_cq_read will keep returning -FI_EAVAIL so just bail out and try again later.
+		 */
+		return 0;
+	} else if (OFI_UNLIKELY(ret < 0)) {
+		NCCL_OFI_WARN("Unable to read from fi_cq_readerr. RC: %d. Error: %s",
+			      ret, fi_strerror(-ret));
+		goto exit;
+	}
 
-static int post_flush_req(nccl_net_ofi_rdma_req_t *req);
+	if (err_entry.flags & FI_REMOTE_WRITE) {
+		req = get_req_from_imm_data(ep, err_entry.data);
+		if (!req) {
+			NCCL_OFI_WARN("Unknown remote write error, could not get CQ data");
+			ret = -EIO;
+			goto exit;
+		}
+	} else {
+		/* For all other operations, ctx should be a req */
+		if (!err_entry.op_context) {
+			NCCL_OFI_WARN("Operation with NULL context completed with error");
+			ret = -EIO;
+			goto exit;
+		}
+		req = err_entry.op_context;
+	}
 
-static int post_eager_copy(nccl_net_ofi_rdma_req_t *req);
+	NCCL_OFI_WARN("Request %p completed with error. RC: %d. Error: %s. Completed length: %ld, Request: %s",
+		      req, err_entry.err,
+		      fi_cq_strerror(rail->cq, err_entry.prov_errno, err_entry.err_data, NULL, 0),
+		      (long)err_entry.len, nccl_net_ofi_req_str(req));
+	if (req->type == NCCL_OFI_RDMA_BOUNCE) {
+		/* A bounce buffer receive failed -- this is an internal error so bail out */
+		NCCL_OFI_WARN("Fatal: Bounce buffer recv completed with error");
+	} else {
+		/* Move user-facing request to error state */
+		set_request_state_to_error(req);
+	}
+
+	/*
+	 * Libfabric error codes directly map to ISO C errno values for standard
+	 * error codes up to FI_ERRNO_OFFSET, and libfabric-specific error codes
+	 * beyond. nccl_net_ofi_retval_translate() will figure out
+	 * how to deal with these, so it is safe to pass up the err as-is.
+	 * However, any special-handling for prov_errno should be handled here.
+	 */
+	ret = err_entry.err;
```

Comment:
Just verifying that `err_entry.err` will be a negative errno value? That's what the retval_translate function needs.

Reply:
the ret from the call itself is negative but the err field is positive. `nccl_net_ofi_retval_translate` still did what it was supposed to via the fallthrough case, but I've fixed it now.